lib/token.cpp

```diff
@@ -1997,8 +1997,8 @@ static bool removeContradiction(std::list<ValueFlow::Value>& values)
                 auto compare = [](const ValueFlow::Value& x, const ValueFlow::Value& y) {
                     return x.compareValue(y, less{});
                 };
-                const ValueFlow::Value& maxValue = std::max(x, y, compare);
-                const ValueFlow::Value& minValue = std::min(x, y, compare);
+                ValueFlow::Value maxValue = std::max(x, y, compare);
+                ValueFlow::Value minValue = std::min(x, y, compare);
```

Comment:
Ticket to improve the checker to detect this issue: https://trac.cppcheck.net/ticket/11086

Reply:
We should implement 11086 for sure!! This is very sneaky problem.

Reply:
however I don't understand why the x and y parameters would be temporary. I assume the compare function returns a bool..

Reply:
compare is returning a bool. however that is just the selector. lets assume that x < y, then minValue will be a *reference* to x, and maxValue a *reference* to y. 

in the code below, we call 'values.remove(minValue)'. values.remove() iterates over all of the values, comparing with operator==() and erasing *every* element that matches the value. so when it found minValue, minValue is erased and as it is a *reference* it is now a *dangling* reference, and the loop over the next element will execute operator== on a dangling reference. So the code is unsafe whenever  minValue isn't the last element in values. 

The simplest fix is to keep it as value rather than a reference. A slightly more optimized implementation would be to rewrite remove() into erase(). I have a patch for that as well but it is less straight forward.

Reply:
a related issue is mentioned in https://en.cppreference.com/w/cpp/algorithm/max . in general, using references for std::min/max as a return value is not a great idea.